### PR TITLE
[KHWEBA-18] add accounts

### PR DESCRIPTION
**Pull request by OlgaOrlova**
Extracted a component - AccountsComponent
Account storage implementation
Update account's balance value in drawer

### CODE REVIEW

**Comment by gift-a** [comment]
how do you send your accounts data to add transactions dialog?

File: src/components/app/app.js [hunk 1]
```diff
@@ -6,6 +6,11 @@ import { TransactionsComponent } from "../transactions/transactions";
 import { AddAccountComponent } from "../add-account-dialog/add-account-dialog";
 import { AddTagComponent } from "../add-tag-dialog/add-tag-dialog";
 import { AboutComponent } from "../about-dialog/about-dialog";
+import { PieChartComponent } from "../pie-chart/pie-chart";
+import { BarChartComponent } from "../bar-chart/bar-chart";
+import { TableTransactionsComponent } from "../table-transactions/table-transactions";
+import { StorageService } from "../../services/storage";
```
Comment (gift-a) [inline src/components/app/app.js]: you have special accounts component. you can use service there. maybe it should be your own service, that works with StoragService

File: src/components/app/app.js [hunk 2]
```diff
@@ -6,6 +6,11 @@ import { TransactionsComponent } from "../transactions/transactions";
...
+import { Account } from "../../services/model/account";
```
Comment (gift-a) [inline src/components/app/app.js]: why do you import Account in App, Add account dialog, Accounts and Drawer? (4 times) Wouldn't it be better to work with it in one place?

File: src/components/app/app.js [hunk 3]
```diff
@@ -6,6 +6,11 @@ import { TransactionsComponent } from "../transactions/transactions";
 import { AddAccountComponent } from "../add-account-dialog/add-account-dialog";
 import { AddTagComponent } from "../add-tag-dialog/add-tag-dialog";
 import { AboutComponent } from "../about-dialog/about-dialog";
+import { PieChartComponent } from "../pie-chart/pie-chart";
+import { BarChartComponent } from "../bar-chart/bar-chart";
+import { TableTransactionsComponent } from "../table-transactions/table-transactions";
```
Comment (gift-a) [inline src/components/app/app.js]: you don't need charts and table transaction imports. they are in transactions component

File: src/components/app/app.js [hunk 4]
```diff
@@ -35,23 +40,55 @@ export class AppComponent {
...
+          new Account("Payoneer", "", "USD")
+        ];
+        StorageService.set("accounts", accounts);
+      }
+      this.initDrawer(accounts);
+    });
+  }
+
+  initDrawer(accounts) {
+    this.drawerComponent = new DrawerComponent(this.drawerMountPoint, {
```
Comment (gift-a) [inline src/components/app/app.js]: do you need to redraw whole drawer? in my opinion it is better to refresh only accounts list

File: src/components/add-transaction-dialog/add-transaction-dialog.html [hunk 5]
```diff
@@ -53,7 +53,7 @@ <h3 class="mdc-typography--subheading2">Type of transaction: </h3>
               <ul class="mdc-list mdc-menu__items">
                 <% for(let i=0; i<accounts.length; i++) {%>
                   <li class="mdc-list-item" role="option" tabindex="0">
-                    <%= accounts[i] %>
+                    <%= accounts[i].name %>
```
Comment (gift-a) [inline src/components/add-transaction-dialog/add-transaction-dialog.html]: can you send not only name, but currency and type? Select provides "string" value. So it should be in format that we can parse easily

**Comment by gift-a** [comment]
where did tag buttons disappear?

File: src/components/app/app.js [hunk 6]
```diff
@@ -35,23 +40,60 @@ export class AppComponent {
       onAboutClick: this.handleAboutOnclick.bind(this)
     });
     this.toolBarComponent.mount();
-    this.drawerComponent = new DrawerComponent(this.drawerMountPoint, {
-      onAddAccountClick: this.handleAddAccountClick.bind(this),
-      onAddTagClick: this.handleAddTagOnclick.bind(this)
+    this.addAccountDialog = new AddAccountComponent(this.addAccountMountPoint, {
```
Comment (gift-a) [inline src/components/app/app.js]: because you have your account component maybe it is better to move add account dialog from here?

File: src/components/add-account-dialog/add-account-dialog.js [hunk 7]
```diff
@@ -51,7 +68,7 @@ export class AddAccountComponent {
   mount() {
     this.mountPoint.innerHTML = template({
       types: ["checking", "savings", "credit card", "cash"],
-      currencies: ["Ukraine, UAH", "United States, USD", "Europe, EUR"]
+      currencies: ["UAH", "USD", "EUR"]
```
Comment (gift-a) [inline src/components/add-account-dialog/add-account-dialog.js]: as I understand you should have changed it to currency symbols

File: src/components/app/app.js [hunk 8]
```diff
@@ -35,23 +40,60 @@ export class AppComponent {
...
+      { accounts: accounts }
+    );
+    this.transactionsComponent.mount();
+  }
+
+  getCurrentAccounts() {
+    StorageService.get("accounts").then(accounts => {
+      if (!accounts) {
+        accounts = [
+          new Account("Privat", "", "UAH"),
```
Comment (ihor7016) [inline src/components/app/app.js]: by default there should be no accounts

File: src/components/app/app.js [hunk 9]
```diff
@@ -35,23 +40,60 @@ export class AppComponent {
...
+          new Account("Payoneer", "", "USD")
+        ];
+        StorageService.set("accounts", accounts);
+      }
+      this.initDrawer(accounts);
```
Comment (ihor7016) [inline src/components/app/app.js]: you should init drawer at component start, and the update it when new data comes

File: src/components/app/app.js [hunk 10]
```diff
@@ -35,23 +40,60 @@ export class AppComponent {
...
+    });
+    this.drawerComponent.mount();
+  }
+
+  handleAddAccountConfirmed(account) {
+    StorageService.get("accounts").then(accounts => {
+      if (!accounts) {
+        StorageService.set("accounts", [account]);
+      } else {
+        accounts.unshift(account);
```
Comment (ihor7016) [inline src/components/app/app.js]: do not mutate --- better use `concat`

**Review by ihor7016 — COMMENTED** [review]
dialog is not being cleared after first transaction add

File: src/components/app/app.js (hunk 2 above)
Comment (gift-a) [reply src/components/app/app.js]: cancel this question

File: src/components/add-transaction-dialog/add-transaction-dialog.html (hunk 5 above)
Comment (gift-a) [reply src/components/add-transaction-dialog/add-transaction-dialog.html]: cancel this question

File: src/components/accounts/accounts-component/accounts-component.js [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+      "click",
+      this.handleAddAccountClick.bind(this)
+    );
+  }
+
+  addAccount(account) {
+    let element = document.createElement("div");
+    element.innerHTML = accountItemTemplate({ account: account });
+    let childNode = this.accList.childNodes[2];
+    this.accList.insertBefore(element.firstChild, childNode);
```
Comment (ihor7016) [inline src/components/accounts/accounts-component/accounts-component.js]: maybe `this.accList.innerHTML = accountItemTemplate({ account: account }) + this.accList.innerHTML`

File: src/components/accounts/add-account-dialog/add-account-dialog.js [hunk 12]
```diff
@@ -15,29 +17,44 @@ export class AddAccountComponent {
 
   handleOk() {
     console.log("accepted");
```
Comment (ihor7016) [inline src/components/accounts/add-account-dialog/add-account-dialog.js]: please remove unneded `console.log`s

File: src/components/accounts/add-account-dialog/add-account-dialog.js [hunk 13]
```diff
@@ -15,29 +17,44 @@ export class AddAccountComponent {
...
+    let account = new Account(
```
Comment (ihor7016) [inline src/components/accounts/add-account-dialog/add-account-dialog.js]: if you're not going to reassign variable -- use const instead of let

File: src/components/app/app.js [hunk 14]
```diff
@@ -35,23 +41,40 @@ export class AppComponent {
...
+  mountTransactionsComponent(accounts) {
+    this.transactionsComponent = new TransactionsComponent(
+      this.transactionsMountPoint,
+      { accounts: accounts }
+    );
+    this.transactionsComponent.mount();
+  }
+
+  getCurrentAccounts() {
+    StorageService.get("accounts").then(accounts => {
```
Comment (ihor7016) [inline src/components/app/app.js]: please move this logic to accounts component (if you need)

File: src/components/app/app.js [hunk 15]
```diff
@@ -35,23 +41,40 @@ export class AppComponent {
...
+      }
+      this.drawerComponent.initAccounts(accounts);
+      this.mountTransactionsComponent(accounts);
+    });
+  }
+
+  initDrawer() {
+    this.drawerComponent = new DrawerComponent(this.drawerMountPoint, {
+      onAddTagClick: this.handleAddTagOnclick.bind(this),
+      addAccountMountPoint: this.addAccountMountPoint
```
Comment (ihor7016) [inline src/components/app/app.js]: please do not pass dom elements to child components

File: src/components/app/app.js [hunk 16]
```diff
@@ -35,23 +41,40 @@ export class AppComponent {
...
     this.addTagDialog = new AddTagComponent(this.addTagMountPoint);
     this.addTagDialog.mount();
     this.aboutDialog = new AboutComponent(this.aboutMountPoint);
     this.aboutDialog.mount();
   }
 
+  mountTransactionsComponent(accounts) {
+    this.transactionsComponent = new TransactionsComponent(
+      this.transactionsMountPoint,
+      { accounts: accounts }
```
Comment (ihor7016) [inline src/components/app/app.js]: you do not need to pass accounts to transactions component -- it is work of Kseniia

File: src/components/accounts/accounts-component/accounts-component.js [hunk 17]
```diff
@@ -0,0 +1,93 @@
...
+    this.addAccountComponent = new AddAccountComponent(
+      this.props.addAccountMountPoint,
+      {
+        onAddAccountConfirmed: this.handleAddAccountConfirmed.bind(this)
+      }
+    );
+  }
+
+  handleAddAccountConfirmed(account) {
+    AccountService.get().then(accounts => {
```
Comment (ihor7016) [inline src/components/accounts/accounts-component/accounts-component.js]: please in `AccountService` create one more method called `add` -- and move this logic there

**Comment by ihor7016** [comment]
image

where 3dot buttons in the tag list?

**Comment by ihor7016** [comment]
image

we need one more field in this dialog -- initial balance -- please add this input to the dialog to the right of the account name

File: src/components/accounts/accounts-component/accounts-component.js (hunk 11 above)
Comment (OlgaOrlova) [reply src/components/accounts/accounts-component/accounts-component.js]: we have to take into account first item too, that's why I've made like this

File: src/components/app/app.js [hunk 18]
```diff
@@ -3,9 +3,14 @@ import template from "./app.html";
 import { DrawerComponent } from "../drawer/drawer";
 import { ToolbarComponent } from "../toolbar/toolbar";
 import { TransactionsComponent } from "../transactions/transactions";
-import { AddAccountComponent } from "../add-account-dialog/add-account-dialog";
+import { AddAccountComponent } from "../accounts/add-account-dialog/add-account-dialog";
 import { AddTagComponent } from "../add-tag-dialog/add-tag-dialog";
 import { AboutComponent } from "../about-dialog/about-dialog";
+import { PieChartComponent } from "../pie-chart/pie-chart";
```
Comment (gift-a) [inline src/components/app/app.js]: you have charts and table in app again. please, remove them. they are in transaction component

File: src/components/app/app.js [hunk 19]
```diff
@@ -3,9 +3,14 @@ import template from "./app.html";
...
+import { BarChartComponent } from "../bar-chart/bar-chart";
+import { TableTransactionsComponent } from "../table-transactions/table-transactions";
+import { StorageService } from "../../services/storage";
```
Comment (gift-a) [inline src/components/app/app.js]: maybe you forgot to remove storage import here. we use own storage service

File: src/components/app/app.js [hunk 20]
```diff
@@ -35,21 +40,31 @@ export class AppComponent {
...
     this.toolBarComponent.mount();
-    this.drawerComponent = new DrawerComponent(this.drawerMountPoint, {
-      onAddAccountClick: this.handleAddAccountClick.bind(this),
-      onAddTagClick: this.handleAddTagOnclick.bind(this)
-    });
-    this.drawerComponent.mount();
-    this.transactionsComponent = new TransactionsComponent(
-      this.transactionsMountPoint
-    );
-    this.transactionsComponent.mount();
```
Comment (gift-a) [inline src/components/app/app.js]: you mustn't change rows 43-46. Your task don't affect on this component

File: src/components/app/app.js [hunk 21]
```diff
@@ -35,21 +40,31 @@ export class AppComponent {
...
     this.addTagDialog.mount();
     this.aboutDialog = new AboutComponent(this.aboutMountPoint);
     this.aboutDialog.mount();
+    this.mountTransactionsComponent();
+  }
+
+  mountTransactionsComponent(accounts) {
+    this.transactionsComponent = new TransactionsComponent(
+      this.transactionsMountPoint,
+      { accounts: accounts }
```
Comment (gift-a) [inline src/components/app/app.js]: you don't need to send accounts to this component, so you don't need this method. please, restore rows 43-46

File: src/components/transactions/transactions.js [hunk 22]
```diff
@@ -46,7 +46,8 @@ export class TransactionsComponent {
     this.barChartComponent = new BarChartComponent(this.barChartMountPoint);
     this.barChartComponent.mount();
     this.addTransactionDialog = new AddTransactionComponent(
-      this.addTransactionDialogMountPoint
+      this.addTransactionDialogMountPoint,
+      { accounts: this.props.accounts }
```
Comment (gift-a) [inline src/components/transactions/transactions.js]: your task doesn't affect on this component. please, remove these changes

File: src/index.scss [hunk 23]
```diff
@@ -24,7 +24,8 @@
 @import "./components/toolbar/toolbar";
 @import "./components/transactions/transactions";
 @import "./components/add-transaction-dialog/add-transaction-dialog";
-@import "./components/add-account-dialog/add-account-dialog";
+@import "components/accounts/add-account-dialog/add-account-dialog";
```
Comment (gift-a) [inline src/index.scss]: add pls "./" in the beginning

**Comment by gift-a** [comment]
storage doesn't save balance
image

**Comment by gift-a** [comment]
tag buttons are broken
image

**Comment by gift-a** [comment]
the add transactions dialog is broken
image

File: src/components/add-transaction-dialog/add-transaction-dialog.js [hunk 24]
```diff
@@ -67,7 +67,7 @@ export class AddTransactionComponent {
 
   mount() {
     this.mountPoint.innerHTML = template({
-      accounts: ["Private", "Cash"],
+      accounts: this.props.accounts,
```
Comment (gift-a) [inline src/components/add-transaction-dialog/add-transaction-dialog.js]: you should not change this file. your task doesn't affect on this component

**Comment by OlgaOrlova** [comment]
Tag buttons aren't  broken, it's doesn't visible, because i haven't got TagComponent

**Comment by gift-a** [comment]
tag buttons are in master. so they should be in your code too

File: src/services/account-service.js [hunk 25]
```diff
@@ -0,0 +1,22 @@
+import { StorageService } from "./storage";
+
+export class AccountService {
+  static add(account) {
+    return AccountService.get().then(accounts => {
```
Comment (ihor7016) [inline src/services/account-service.js]: `this.get()`

File: src/services/account-service.js [hunk 26]
```diff
@@ -0,0 +1,22 @@
...
+      if (!accounts) {
+        AccountService.set([account]);
+      } else {
+        let updatedAccounts = [account].concat(accounts);
+        AccountService.set(updatedAccounts);
```
Comment (ihor7016) [inline src/services/account-service.js]: `this.set`

Comment (ihor7016) [reply src/services/account-service.js]: also you could just do `this.set([account, ...accounts])` -- es6 array rest-spread

File: src/components/drawer/drawer.js [hunk 27]
```diff
@@ -1,24 +1,24 @@
 import template from "./drawer.html";
 import { MDCPersistentDrawer } from "@material/drawer";
 import { ButtonMoreComponent } from "../button-more/button-more";
+import { AccountsComponent } from "../accounts/accounts-component/accounts-component";
 
 export class DrawerComponent {
-  constructor(mountPoint, props) {
+  constructor(mountPoint, dialogMountPoint, props) {
```
Comment (ihor7016) [inline src/components/drawer/drawer.js]: you do not need to pass any mount points except -- one first `mountPoint` for component itself. You can mount dialog anywhere in the app, in any component -- it does not hurt its rendering

File: src/components/drawer/drawer.js [hunk 28]
```diff
@@ -28,7 +28,26 @@ export class DrawerComponent {
     this.drawer = new MDCPersistentDrawer(this.drawerRoot);
   }
 
+  initAccounts(accounts) {
+    accounts.forEach(item => {
```
Comment (ihor7016) [inline src/components/drawer/drawer.js]: please move this to accounts component

File: src/components/drawer/drawer.js [hunk 29]
```diff
@@ -37,12 +56,15 @@ export class DrawerComponent {
       }).mount();
     });
   }
+  onAddAccountConfirmed(account) {
+    this.addAccount(account);
```
Comment (ihor7016) [inline src/components/drawer/drawer.js]: move this to accounts component

File: src/components/drawer/drawer.js [hunk 30]
```diff
@@ -37,12 +56,15 @@ export class DrawerComponent {
...
+  }
+
+  addAccount(account) {
+    this.accountsComponent.addAccountToHead(account);
```
Comment (ihor7016) [inline src/components/drawer/drawer.js]: move this to accounts component

File: src/components/drawer/drawer.js (hunk 27 above)
Comment (ihor7016) [reply src/components/drawer/drawer.js]: Actually you should move dialog mounting to accounts component

File: src/components/accounts/add-account-dialog/add-account-dialog.html [hunk 31]
```diff
@@ -34,12 +44,14 @@ <h2 class="mdc-dialog__header__title">
...
               <div class="mdc-select__label">Currency</div>
-              <div class="mdc-select__selected-text"></div>
+              <div class="add-account-dialog__currency-text mdc-select__selected-text"></div>
               <div class="mdc-select__bottom-line"></div>
             </div>
             <div class="mdc-menu mdc-select__menu add-account-dialog__select-menu">
               <ul class="mdc-list mdc-menu__items">
-                <% for(let i=0; i<currencies.length; i++) {%>
+                <% for(let i=0; i
+                <currencies.length
```
Comment (ihor7016) [inline src/components/accounts/add-account-dialog/add-account-dialog.html]: please fix formatting here

File: src/components/accounts/accounts-component/accounts-component.js [hunk 32]
```diff
@@ -0,0 +1,111 @@
+import template from "./accounts-component.html";
+import accountItemTemplate from "./account-item.html";
+import { ButtonMoreComponent } from "../../button-more/button-more";
+import { AddAccountComponent } from "../add-account-dialog/add-account-dialog";
+import { AccountService } from "../../../services/account-service";
+
+export class AccountsComponent {
+  constructor(mountPoint, addAccountMountPoint, props) {
```
Comment (ihor7016) [inline src/components/accounts/accounts-component/accounts-component.js]: why are you passing `addAccountMountPoint ` ?

Comment (ihor7016) [reply src/components/accounts/accounts-component/accounts-component.js]: you should not pass any mount points except one for component

File: src/components/accounts/accounts-component/accounts-component.js [hunk 33]
```diff
@@ -0,0 +1,111 @@
...
+      "click",
+      this.handleAddAccountClick.bind(this)
+    );
+  }
+
+  addAccountToHead(account) {
+    let element = document.createElement("div");
+    element.innerHTML = accountItemTemplate({ account: account });
+    let childNode = this.accList.childNodes[2];
+    this.accList.insertBefore(element.firstChild, childNode);
```
Comment (ihor7016) [inline src/components/accounts/accounts-component/accounts-component.js]: `this.accList.innerHTML = accountItemTemplate({ account: account }) + this.accList.innerHTML`

File: src/components/accounts/accounts-component/accounts-component.js [hunk 34]
```diff
@@ -0,0 +1,111 @@
...
+    this.initMoreBtns();
+  }
+
+  addAccount(account) {
+    let element = document.createElement("div");
```
Comment (ihor7016) [inline src/components/accounts/accounts-component/accounts-component.js]: please do not create elements with js -- use innerHTML

File: src/components/accounts/accounts-component/accounts-component.js [hunk 35]
```diff
@@ -0,0 +1,111 @@
...
+    element.innerHTML = accountItemTemplate({ account: account });
+    this.accList.appendChild(element.firstChild);
+    this.initMoreBtns();
+  }
+
+  handleAddAccountClick() {
+    this.addAccountComponent.mount();
```
Comment (ihor7016) [inline src/components/accounts/accounts-component/accounts-component.js]: why are you mounting component on click? you should do it in `mountChildren`

File: src/components/drawer/drawer.js (hunk 27 above)
Comment (OlgaOrlova) [reply src/components/drawer/drawer.js]: @ihor7016 How should we find mount point for inner component? We pass dialog mount point because it corresponds to app.html. If I remove passing by constructor - we have to find this element with document.querySelector , which is also bad

File: src/components/accounts/accounts-component/accounts-component.js (hunk 35 above)
Comment (OlgaOrlova) [reply src/components/accounts/accounts-component/accounts-component.js]: because it's easier to get new empty dialog every time, instead of clearing all the fields and selections when it's hiding

**Review by Konstanty-Ivashchenko — DISMISSED** [review]

**Review by Vladyslav-Plakhuta — DISMISSED** [review]

File: src/components/accounts/accounts-component/accounts-component.js [hunk 36]
```diff
@@ -0,0 +1,105 @@
...
+  }
+
+  initMoreBtns() {
+    this.moreBtnMountPoints = this.mountPoint.querySelectorAll(
+      ".account__more-button"
+    );
+    Array.from(this.moreBtnMountPoints).forEach(point => {
+      new ButtonMoreComponent(point, {
+        position: "right",
+        onDeleteClicked: this.handleDeleteClick.bind(this),
```
Comment (gift-a) [inline src/components/accounts/accounts-component/accounts-component.js]: decided to use onDeleteClick instead Clicked

File: src/components/accounts/add-account-dialog/add-account-dialog.js [hunk 37]
```diff
@@ -48,10 +57,30 @@ export class AddAccountComponent {
     this.dialog.listen("MDCDialog:cancel", this.handleCancel.bind(this));
   }
 
+  clean() {
+    this.accountNameInput.value = "";
+    this.balanceNameInput.value = "";
```
Comment (gift-a) [inline src/components/accounts/add-account-dialog/add-account-dialog.js]: you can clean selects with `this.select.selectedIndex = -1;`

File: src/services/account-service.js [hunk 38]
```diff
@@ -1,6 +1,40 @@
...
+          });
+        }
+      })
+      .then(account => {
+        this.get().then(accounts => {
+          let index = accounts.findIndex(item => {
+            return item.name === account.name;
+          });
+          account.balance = parseInt(account.balance) + amount;
+          accounts[index] = account;
```
Comment (gift-a) [inline src/services/account-service.js]: first time you save balance in storage as `string`. After update you save it as `number`, and don't need parse it as integer every time you update balance. so I think it is better to save it in storage as `number` by default:
image
and again. "confirm" instead "confirmed"

File: src/components/accounts/accounts-component/accounts-component.html [hunk 39]
```diff
@@ -0,0 +1,10 @@
+<div class="account">
```
Comment (ihor7016) [inline src/components/accounts/accounts-component/accounts-component.html]: if component called `accounts` why class name is `account` ?

Comment (ihor7016) [inline src/components/accounts/accounts-component/accounts-component.html]: we also do not use `component` in file names -- so it should be `accounts.html`

File: src/components/accounts/accounts-component/accounts-component.js [hunk 40]
```diff
@@ -0,0 +1,105 @@
+import template from "./accounts-component.html";
```
Comment (ihor7016) [inline src/components/accounts/accounts-component/accounts-component.js]: we do not use `component` in folder or file names

File: src/components/accounts/accounts-component/accounts-component.js [hunk 41]
```diff
@@ -0,0 +1,105 @@
...
+      this.addAccount(item);
+    });
+  }
+
+  initData() {
+    this.accountsList.innerHTML = "";
+    AccountListService.get().then(accounts => {
+      if (!accounts) {
+        accounts = [];
+        AccountListService.set(accounts);
```
Comment (ihor7016) [inline src/components/accounts/accounts-component/accounts-component.js]: please do not set data here -- it would be better if `AccountListService` would return empty list if there is no items there or just do like this

```js
AccountListService.get().then(accounts => this.initAccounts(accounts || []) })
```

File: src/components/accounts/accounts-component/accounts-component.js [hunk 42]
```diff
@@ -0,0 +1,105 @@
...
+      }
+      this.initAccounts(accounts);
+    });
+  }
+
+  updateAccountData(transaction) {
+    AccountListService.update(
+      transaction.account,
+      parseInt(transaction.type + transaction.amount)
+    ).then(_ => {
```
Comment (ihor7016) [inline src/components/accounts/accounts-component/accounts-component.js]: why do you need `_` here ?? if you do not need parameters you do like `() => this.initData()`

**Comment by ihor7016** [comment]
please add all these types of accounts to the dialog


image

File: src/services/account-service.js [hunk 43]
```diff
@@ -1,6 +1,40 @@
...
+        if (!accounts) {
+          return new Error("empty accounts");
+        } else {
+          return accounts.find(item => {
+            return item.name === account.name;
+          });
+        }
+      })
+      .then(account => {
+        this.get().then(accounts => {
```
Comment (ihor7016) [inline src/services/account-service.js]: why are you doing 2 `get()`s here?

File: src/components/accounts/add-account-dialog/add-account-dialog.js [hunk 44]
```diff
@@ -0,0 +1,103 @@
...
+      name: parseInt(this.accountNameInput.value),
+      balance: parseInt(this.balanceNameInput.value),
+      type: this.accountType.innerText,
+      currency: this.accountCurrency.innerText
+    });
+    this.clean();
+  }
+
+  handleCancel() {
+    console.log("declined");
```
Comment (ihor7016) [inline src/components/accounts/add-account-dialog/add-account-dialog.js]: please remove this console.log

File: src/components/accounts/accounts-component/accounts-component.js [hunk 45]
```diff
@@ -0,0 +1,104 @@
+import template from "./accounts.html";
```
Comment (ihor7016) [inline src/components/accounts/accounts-component/accounts-component.js]: we do not use `component` in naming files/directories

File: src/services/account-service.js [hunk 46]
```diff
@@ -1,11 +1,42 @@
...
+        let updatedAccounts = [account].concat(accounts);
+        this.set(updatedAccounts);
+      }
+    });
+  }
+
+  static update(transactionAccount, amount) {
+    return this.get()
+      .then(accounts => this.initAccounts(accounts || []))
+      .then(accounts => {
```
Comment (ihor7016) [inline src/services/account-service.js]: please review this code -- not sure that it works

File: src/services/account-service.js [hunk 47]
```diff
@@ -1,11 +1,42 @@
...
   static get() {
     return StorageService.get("accountList");
   }
 
   static set(value) {
     return StorageService.set("accountList", value);
   }
+
+  static initAccounts(accounts) {
+    this.set(accounts);
```
Comment (ihor7016) [inline src/services/account-service.js]: you do not need this method

File: src/index.scss [hunk 48]
```diff
@@ -24,9 +24,10 @@
 @import "./components/toolbar/toolbar";
 @import "./components/transactions/transactions";
 @import "./components/add-transaction-dialog/add-transaction-dialog";
-@import "./components/add-account-dialog/add-account-dialog";
+@import "components/add-account-dialog/add-account-dialog";
```
Comment (ihor7016) [inline src/index.scss]: why did you remove `./` from here?

File: src/services/account-service.js [hunk 49]
```diff
@@ -1,6 +1,35 @@
...
+  static update(transactionAccount, amount) {
+    return this.get()
+      .then(accounts => {
+        return accounts || [];
+      })
+      .then(accounts => {
+        let account = accounts.find(item => {
+          return item.name === transactionAccount.name;
+        });
+        let index = accounts.findIndex(item => {
```
Comment (ihor7016) [inline src/services/account-service.js]: why do you need 2 searches -- please leave one by index

File: src/components/app/app.js [hunk 50]
```diff
@@ -32,17 +29,43 @@ export class AppComponent {
...
-      onAddAccountClick: this.handleAddAccountClick.bind(this),
       onAddTagClick: this.handleAddTagOnclick.bind(this)
     });
     this.drawerComponent.mount();
-    this.transactionsComponent = new TransactionsComponent(
-      this.transactionsMountPoint
+  }
+
+  initAccountComponent() {
+    this.accountsComponent = new AccountsComponent(
```
Comment (ihor7016) [inline src/components/app/app.js]: why do you need accounts component here? it should be only in drawer

File: src/components/accounts/accounts.js [hunk 51]
```diff
@@ -0,0 +1,119 @@
...
+import { AddAccountDialogComponent } from "../add-account-dialog/add-account-dialog";
+
+export class AccountsComponent {
+  constructor(mountPoint, props) {
+    this.mountPoint = mountPoint;
+    this.props = props;
+  }
+
+  querySelectors() {
+    this.accountsRoot = this.mountPoint.querySelector(".accounts");
```
Comment (gift-a) [inline src/components/accounts/accounts.js:14]: maybe  `mount point`, not `root`?

File: src/components/accounts/accounts.js [hunk 52]
```diff
@@ -0,0 +1,119 @@
...
+    this.addAccountButton = this.mountPoint.querySelector(
+      ".accounts__add-account-dialog-activation"
+    );
+    this.addAccountMountPoint = document.querySelector(
+      ".accounts__add-account-dialog"
+    );
+    this.accountsList = this.mountPoint.querySelector(".accounts__list-items");
+  }
+
+  handleAddAccountConfirmed(account) {
```
Comment (gift-a) [inline src/components/accounts/accounts.js:27]: we decided to name as `confirm`, not `confirmed`

File: src/components/accounts/accounts.js [hunk 53]
```diff
@@ -0,0 +1,119 @@
...
+    AccountListService.add(account);
+    this.addAccountToHead(account);
+  }
+
+  handleAddAccountClicked() {
+    this.addAccountDialogComponent.showDialog();
+  }
+
+  initMoreBtns() {
+    this.moreBtnMountPoints = this.mountPoint.querySelectorAll(
```
Comment (gift-a) [inline src/components/accounts/accounts.js:37]: you do `this.moreBtnMountPoints = this.mountPoint.querySelectorAll` twice. here and on the row 15. is it "ok" or "typo"?

File: src/components/accounts/accounts.js [hunk 54]
```diff
@@ -0,0 +1,119 @@
...
+    );
+    this.accountsList = this.mountPoint.querySelector(".accounts__list-items");
+  }
+
+  handleAddAccountConfirmed(account) {
+    AccountListService.add(account);
+    this.addAccountToHead(account);
+  }
+
+  handleAddAccountClicked() {
```
Comment (gift-a) [inline src/components/accounts/accounts.js:32]: click

File: src/components/app/app.js [hunk 55]
```diff
@@ -32,27 +28,31 @@ export class AppComponent {
       onMenuClick: this.handleToolbarMenuClick.bind(this)
     });
     this.toolBarComponent.mount();
+    this.initDrawer();
+    this.addTagDialog = new AddTagComponent(this.addTagMountPoint);
+    this.addTagDialog.mount();
+    this.transactionsComponent = new TransactionsComponent(
+      this.transactionsMountPoint,
+      {
+        onTransactionAdded: this.handleTransactionAdded.bind(this)
```
Comment (gift-a) [inline src/components/app/app.js:37]: add, without 'ed'

File: src/components/drawer/drawer.js [hunk 56]
```diff
@@ -11,24 +12,33 @@ export class DrawerComponent {
...
+    this.accountsComponent = new AccountsComponent(this.accountsMountPoint);
+    this.accountsComponent.mount();
+  }
 
   initMDC() {
     this.drawer = new MDCPersistentDrawer(this.drawerRoot);
   }
 
   initMoreBtns() {
+    this.moreBtnMountPoints = this.mountPoint.querySelectorAll(
```
Comment (gift-a) [inline src/components/drawer/drawer.js:39]: the same. you do `this.moreBtnMountPoints = this.mountPoint.querySelectorAll` twice

File: src/components/transactions/transactions.js [hunk 57]
```diff
@@ -44,6 +44,7 @@ export class TransactionsComponent {
     this.barChartComponent.update(data);
     this.pieChartComponent.update(data);
     this.addStoredData(data);
+    this.props.onTransactionAdded(data);
```
Comment (gift-a) [inline src/components/transactions/transactions.js:47]: add

**Review by ihor7016 — APPROVED** [review]
